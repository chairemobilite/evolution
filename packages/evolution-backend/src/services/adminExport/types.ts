/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

type ExportOptions = {
    /** Specifies which set of responses should be exported. 'validated' means
     * it exports the validated data if available, 'participant' is the original
     * participant responses */
    responseType: 'participant' | 'validatedIfAvailable';
};
