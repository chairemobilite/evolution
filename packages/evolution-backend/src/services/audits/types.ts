/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

import { ExtendedPlaceAttributes } from 'evolution-common/lib/services/baseObjects/Place';
import { ExtendedPersonAttributes } from 'evolution-common/lib/services/baseObjects/Person';
import { ExtendedHouseholdAttributes } from 'evolution-common/lib/services/baseObjects/Household';
import { ExtendedJourneyAttributes } from 'evolution-common/lib/services/baseObjects/Journey';
import { ExtendedSegmentAttributes } from 'evolution-common/lib/services/baseObjects/Segment';
import { ExtendedVisitedPlaceAttributes } from 'evolution-common/lib/services/baseObjects/VisitedPlace';
import { ExtendedTripAttributes } from 'evolution-common/lib/services/baseObjects/Trip';
import { CorrectedResponse } from 'evolution-common/lib/services/questionnaire/types';

/**
 * Parser function type for converting survey response values to proper types
 * before object validation. For example, converting 'yes'/'no' strings to boolean values.
 * Parsers modify the corrected response in place.
 */
export type SurveyObjectParserInterview<TInput> = (attributes: TInput) => void;
export type SurveyObjectParser<TInput, CorrectedResponse> = (
    attributes: TInput,
    correctedResponse: CorrectedResponse
) => void;

/**
 * Configuration for survey object parsers.
 * Each parser takes the corrected response and modifies the response data in place
 * to convert string choice values to proper types before object creation.
 */
export type SurveyObjectParsers = {
    interview?: SurveyObjectParserInterview<CorrectedResponse>;
    household?: SurveyObjectParser<ExtendedHouseholdAttributes, CorrectedResponse>;
    home?: SurveyObjectParser<ExtendedPlaceAttributes, CorrectedResponse>;
    person?: SurveyObjectParser<ExtendedPersonAttributes, CorrectedResponse>;
    journey?: SurveyObjectParser<ExtendedJourneyAttributes, CorrectedResponse>;
    visitedPlace?: SurveyObjectParser<ExtendedVisitedPlaceAttributes, CorrectedResponse>;
    trip?: SurveyObjectParser<ExtendedTripAttributes, CorrectedResponse>;
    segment?: SurveyObjectParser<ExtendedSegmentAttributes, CorrectedResponse>;
};
