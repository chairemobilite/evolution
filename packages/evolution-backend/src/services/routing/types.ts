/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { RoutingOrTransitMode } from 'chaire-lib-common/lib/config/routingModes';
import { SummaryResponse, SummarySuccessResult } from 'chaire-lib-common/lib/api/TrRouting/trRoutingApiV2';

export type RouteCalculationParameter = {
    /**
     * The origin of the route to calculate
     */
    origin: GeoJSON.Point | GeoJSON.Feature<GeoJSON.Point>;
    /**
     * The destination of the route to calculate
     */
    destination: GeoJSON.Point | GeoJSON.Feature<GeoJSON.Point>;
    /**
     * The time of the trip, in seconds since midnight
     */
    departureSecondsSinceMidnight: number;
    /**
     * The date of the trip, in the format 'YYYY-MM-DD'
     */
    departureDateString: string;
    /**
     * If `transit` mode is requested, the scenario to use for calculations
     */
    transitScenario?: string;
};

type RoutingTimeDistanceResult = {
    // The source of the routing calculation. Can be used to identify the method
    // called for this calculation
    source: string;
} & (
    | {
          status: 'no_routing_found';
      }
    | {
          status: 'error';
          error: string;
      }
    | {
          status: 'success';
          distanceM: number;
          travelTimeS: number;
      }
);

export type RoutingTimeDistanceResultByMode = {
    [mode in RoutingOrTransitMode]?: RoutingTimeDistanceResult;
};

export type SummaryResult = {
    // The source of the summary calculation. Can be used to identify the method
    // called for this calculation
    source: string;
} & (
    | {
          status: 'no_routing_found';
      }
    | {
          status: 'error';
          error: string;
      }
    | {
          status: 'success';
          nbRoutes: number;
          lines: SummarySuccessResult['result']['lines'];
      }
);
