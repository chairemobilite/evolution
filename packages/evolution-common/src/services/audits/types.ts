/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

export type Audit = {
    /**
     * Indicate the current version of a validation. If the validation changes, the version will change
     */
    version: number;
    /**
     * Whether this audit message is simply a warning or an error. Defaults to false
     */
    isWarning?: boolean;
    /**
     * The code of this validation. It should be unique per interview object.
     */
    errorCode: string;
    /**
     * A valid i18n key to translate this message. If not set, the errorCode will be used as the i18n key.
     */
    message?: string;
    /**
     * Whether to ignore this validation in the interface. Defaults to false
     */
    ignore?: boolean;
};

export type AuditForObject = Audit & {
    /**
     * The type of the object this audit applies to. Can be 'interview', 'home', 'person', 'household', etc.
     */
    objectType: string;
    /**
     * The object's UUID. If the object is a single object without a specific uuid, like the interview or household, the interview's uuid is used for this field.
     */
    objectUuid: string;
};

export type Audits = { [key: string]: Audit };
