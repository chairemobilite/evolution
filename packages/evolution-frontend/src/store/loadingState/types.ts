/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
export enum LoadingStateActionTypes {
    INCREMENT_LOADING_STATE = 'INCREMENT_LOADING_STATE',
    DECREMENT_LOADING_STATE = 'DECREMENT_LOADING_STATE',
}

export type LoadingStateAction =
    | {
          type: LoadingStateActionTypes.INCREMENT_LOADING_STATE;
          loadingState: number;
      }
    | {
          type: LoadingStateActionTypes.DECREMENT_LOADING_STATE;
          loadingState: number;
      };

export interface LoadingStateState {
    readonly loadingState?: number;
}
