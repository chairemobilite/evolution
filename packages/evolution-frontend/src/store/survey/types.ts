/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
export enum SurveyActionTypes {
    SET_INTERVIEW = 'SET_INTERVIEW',
    UPDATE_INTERVIEW = 'UPDATE_INTERVIEW'
}

export type SurveyAction =
    | {
          type: SurveyActionTypes.SET_INTERVIEW;
          // TODO Properly type the interview
          interview: any;
          interviewLoaded: boolean;
      }
    | {
          type: SurveyActionTypes.UPDATE_INTERVIEW;
          // TODO Properly type the interview
          interview: any;
          interviewLoaded: boolean;
          errors: {
              [key: string]: {
                  [key: string]: string;
              };
          };
          submitted: boolean;
      };

export interface SurveyState {
    readonly interview?: any;
    readonly interviewLoaded?: boolean;
    readonly errors?: {
        [key: string]: {
            [key: string]: string;
        };
    };
    readonly submitted?: boolean;
}
