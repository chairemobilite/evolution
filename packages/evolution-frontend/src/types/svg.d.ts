/*
 * Copyright 2025, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */

// FIXME See if we need to move this declaration to the root of the project to allow other typescript files to use it
declare module '*.svg' {
    const content: any;
    export default content;
}
