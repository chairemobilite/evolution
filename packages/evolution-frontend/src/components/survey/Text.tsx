/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import Markdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { translateString } from 'evolution-common/lib/utils/helpers';
import { WidgetStatus } from 'evolution-common/lib/services/questionnaire/types';
import { TextWidgetConfig } from 'evolution-common/lib/services/questionnaire/types';
import { UserInterviewAttributes } from 'evolution-common/lib/services/questionnaire/types';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';

type TextProps = {
    widgetConfig: TextWidgetConfig;
    interview: UserInterviewAttributes;
    user: CliUser;
    widgetStatus: WidgetStatus;
    path: string;
};

export const Text: React.FunctionComponent<TextProps & WithTranslation> = ({
    widgetConfig,
    widgetStatus,
    i18n,
    interview,
    path,
    user
}) => {
    if (!widgetStatus.isVisible) {
        return null;
    }

    const content = translateString(widgetConfig.text, i18n, interview, path, user);
    if (_isBlank(content)) {
        return null;
    }

    return (
        <div className="survey-section__text">
            {widgetConfig.containsHtml && <div dangerouslySetInnerHTML={{ __html: content as string }} />}
            {!widgetConfig.containsHtml && (
                <Markdown
                    remarkPlugins={[[remarkGfm, { singleTilde: false }]]}
                    className={
                        (widgetConfig.align || 'left') + ' ' + (widgetConfig.classes ? widgetConfig.classes : '')
                    }
                >
                    {content as string}
                </Markdown>
            )}
        </div>
    );
};

export default withTranslation()(Text);
