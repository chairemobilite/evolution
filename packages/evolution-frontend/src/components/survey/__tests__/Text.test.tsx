/*
 * Copyright 2024, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import each from 'jest-each';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations);

import { interviewAttributes } from '../../inputs/__tests__/interviewData.test';
import Text from '../Text';
import { WidgetStatus } from 'evolution-common/lib/services/questionnaire/types';

// Mock react-markdown and remark-gfm as they use syntax not supported by jest
jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
};

const commonWidgetConfig = {
    type: 'text' as const,
    text: 'random text'
};

const defaultWidgetStatus: WidgetStatus = {
    path: 'foo',
    isVisible: true,
    modalIsOpen: false,
    isDisabled: false,
    isCollapsed: false,
    isEmpty: false,
    isCustomEmpty: false,
    isValid: true,
    isResponded: true,
    isCustomResponded: false,
    errorMessage: undefined,
    currentUpdateKey: 1,
    value: null
};

each([
    ['Default values', commonWidgetConfig],
    ['Contains html', { ...commonWidgetConfig, containsHtml: true, text: '<br> bold text</br>blabla' }],
    ['No html', { ...commonWidgetConfig, containsHtml: false, text: '<br>text not to be in bold</br>other' }],
    ['No html, with classes and alignment', { ...commonWidgetConfig, containsHtml: false, classes: 'my-class', align: 'right' }],
]).describe('Text widget: %s', (_widget, widgetConfig) => {

    test('Render widget', () => {

        const wrapper = TestRenderer.create(
            <Text
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Widget accessibility', async () => {
        const { container } = render(
            <Text
                path='home.region'
                widgetConfig={widgetConfig}
                interview={interviewAttributes}
                user={userAttributes}
                widgetStatus={defaultWidgetStatus}
            />
        );
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});

test('Widget invisible, should be null', () => {
    const widgetStatus = _cloneDeep(defaultWidgetStatus);
    widgetStatus.isVisible = false;
    const wrapper = TestRenderer.create(
        <Text
            path='home.region'
            widgetConfig={commonWidgetConfig}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={widgetStatus}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Blank text on widget, should be null', () => {
    const widgetConfig = _cloneDeep(commonWidgetConfig);
    widgetConfig.text = jest.fn().mockReturnValue('') as any;
    const wrapper = TestRenderer.create(
        <Text
            path='home.region'
            widgetConfig={widgetConfig}
            interview={interviewAttributes}
            user={userAttributes}
            widgetStatus={defaultWidgetStatus}
        />
    );
    expect(wrapper).toMatchSnapshot();
});
