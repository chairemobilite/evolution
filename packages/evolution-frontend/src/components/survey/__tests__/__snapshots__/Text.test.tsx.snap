// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blank text on widget, should be null 1`] = `null`;

exports[`Text widget: Contains html Render widget 1`] = `
<div
  className="survey-section__text"
>
  <div
    dangerouslySetInnerHTML={
      {
        "__html": "<br> bold text</br>blabla",
      }
    }
  />
</div>
`;

exports[`Text widget: Default values Render widget 1`] = `
<div
  className="survey-section__text"
>
  <Markdown
    className="left "
    remarkPlugins={
      [
        [
          "remark-gfm",
          {
            "singleTilde": false,
          },
        ],
      ]
    }
  >
    random text
  </Markdown>
</div>
`;

exports[`Text widget: No html Render widget 1`] = `
<div
  className="survey-section__text"
>
  <Markdown
    className="left "
    remarkPlugins={
      [
        [
          "remark-gfm",
          {
            "singleTilde": false,
          },
        ],
      ]
    }
  >
    &lt;br&gt;text not to be in bold&lt;/br&gt;other
  </Markdown>
</div>
`;

exports[`Text widget: No html, with classes and alignment Render widget 1`] = `
<div
  className="survey-section__text"
>
  <Markdown
    className="right my-class"
    remarkPlugins={
      [
        [
          "remark-gfm",
          {
            "singleTilde": false,
          },
        ],
      ]
    }
  >
    random text
  </Markdown>
</div>
`;

exports[`Widget invisible, should be null 1`] = `null`;
